### DISCUSSION TITLE
Population pyramid Stat

### DISCUSSION THREAD

**Discussion by EvaMaeRey** [ggplot2-extenders/ggplot-extension-club#33] [Ideas]
Moving to its own conversation. Originally in  #26 (statexpress)

**Comment by teunbrand** [comment]
I don't really think population pyramids should be a stat and perhaps they should be facets instead.
You could ensure that the left panel has a reverse-transform scale, you could have appropriate spacing for the middle axis and stats automagically compute by facet anyway!

**Reply by EvaMaeRey** [reply]
Sounds like a good fit!  

W/ Stat there was a lot of awkwardness w x scale.  You can write something that helps out, but doesn't feel quite right.    API looked like this... I think you could get cleaner feel combining the rect layer, and text layer, and the x scale.   

```{r}
pop_percent_cat_data %>%
  ggplot() + 
  aes(x = percent,  y = age_cat, pyramid_cat = sex) +
  geom_rect(stat = StatPyramid,  sep = 1) +  
  scale_x_pyramid(sep = 1) +  
  aes(label = age_cat)  + 
  geom_text(stat = StatPyramidlabel) 
```

W/ facet a nice API would be like this I guess:

```{r}
pop_percent_cat_data %>%
  ggplot() + 
  aes(x = percent, 
      y = age_cat) +
 geom_col() +
  facet_pyramid(facet = vars(sex)
```

I'm putting current Stat approach in fresh comment for reference.

**Reply by EvaMaeRey** [reply]
Yeah, especially the automagic in-between y axis spacing sounds pretty great.

**Reply by EvaMaeRey** [reply]
![image](https://github.com/user-attachments/assets/cf699ff8-2fa2-4ae2-8d0d-7e3a141b5e2f)

ggpol::facet_share() as seen in 'a grammar guide'... I forgot all about it.
https://evamaerey.github.io/ggplot2_grammar_guide/facet.html#49   

But like ggbreak::scale_x_cut()-type method has appeal; lets you you can use faceting for other vars (e.g. watch your pyramid change over the years, facet_wrap(~years)))

**Comment by EvaMaeRey** [comment]
``` r
library(tidyverse)

oregon_population_pyramid_data <-
  read_csv("https://raw.githubusercontent.com/rfortherestofus/blog/main/population-pyramid-part-1/oregon_population_pyramid_data.csv")
#> Rows: 1296 Columns: 4
#> ── Column specification ────────────────────────────────────────────────────────
#> Delimiter: ","
#> chr (3): county, age, gender
#> dbl (1): percent
#> 
#> ℹ Use `spec()` to retrieve the full column specification for this data.
#> ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
```

``` r

compute_panel_pyramid <- function(data, scales, sep = 0, neg_cat = NULL){
  
  if(is.null(neg_cat)){neg_cat <- sort(data$pyramid_cat)[1] }
  
  data %>% 
  mutate(split = ifelse(pyramid_cat == neg_cat, -sep, sep)) %>% 
  mutate(xmin = 0 + split) %>% 
  mutate(xmax = x * ifelse(pyramid_cat == neg_cat, -1, 1) + split) %>% 
  mutate(y = as.numeric(as.factor(y))) %>% 
  mutate(ymax = y + .45*1) %>% 
  mutate(ymin = y - .45*1)
  
}

StatPyramid <- ggproto("StatPyramid", 
                       Stat,
                       compute_panel = compute_panel_pyramid,
                       default_aes = aes(fill = after_stat(pyramid_cat)))


compute_pyramid_label <- function(data, scales){
  
  data %>% 
    distinct(y, label) %>% 
    mutate(x = 0)
  
}

StatPyramidlabel <- ggproto("StatPyramidlabel",
                            Stat,
                            compute_panel = compute_pyramid_label)


scale_x_pyramid <- function(sep = 1, breaks = 0:10, ...){
  
    scale_x_continuous(breaks = c(-rev(breaks) - sep, breaks + 1),
                     labels = paste0(c(rev(breaks), breaks), "%"),
                     limits = c(-max(breaks),max(breaks)), ...)
  
}
  
  

oregon_population_pyramid_data %>%
  filter(county == "Baker") %>%
  ggplot() + 
  aes(x = percent*100, 
      y = fct_inorder(age), 
      pyramid_cat = gender) +
  geom_rect(stat = StatPyramid,
            sep = 1) + 
  aes(label = age)  + 
  geom_text(stat = StatPyramidlabel) + 
  scale_x_pyramid() +
  labs(y = NULL, x = NULL, fill = NULL) + 
  theme(axis.text.y = element_blank(),
        axis.ticks.y = element_blank(),
        legend.position = "top",
        legend.justification = "left", 
        panel.grid = element_blank(),
        panel.background = element_rect(fill = "whitesmoke")
        )
```

![](https://i.imgur.com/eyG8M6s.png)

<sup>Created on 2024-08-05 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by EvaMaeRey** [reply]
data and pop pyramid knowledge from @dgkeyes!  see also #26.

**Reply by EvaMaeRey** [reply]
From interpretation perspective, I think it would be cool to have another layer complementing the pyramid - something like geom_pyramid_null() that shows what the pyramid would look like there was equal distribution between MF for each age category.  It could be outline or grey+transparent. When moving from geom_bar(position = "dodge") to pyramid, it feels like it's more effort to say which is bigger.

**Reply by EvaMaeRey** [reply]
Also interesting to think of pop chart like diverging bar charts: https://www.jstatsoft.org/article/view/v057i05

So, I wanted to try making a population pyramid with a diverging bar chart approach from ggstats, which works well! A new position is used.  I did see if removing fill aes or replacing with one category would automatically produce the 'drop shadow' reference layer (if MF were equal), but I didn't get this working - but I'm pretty new to ggstats, so maybe it's possible and I just don't know the details.

``` r
library(tidyverse)
age_cats <- c("18-29", "30-49"," 50-64", "65+")
sex_cats <- c("Male", "Female")

sample40replaceT <- function(x){
  sample(x, size = 40, replace = T) %>% factor(levels = x)
}

survey <- data.frame(c1 = sample40replaceT(age_cats),
                     c2 = sample40replaceT(sex_cats))

survey %>% 
  ggplot() + 
  aes(y = c1, fill = c2) + 
  geom_bar(position = ggstats::position_likert_count())
```

![](https://i.imgur.com/fFQWdrg.png)

``` r


# drop shadow layer (if MF counts were equal) attempt
last_plot() + 
  aes(fill = NULL)
```

![](https://i.imgur.com/6GGeW4I.png)

``` r

# drop shadow layer attempt 2
last_plot() + 
  aes(fill = "All")
```

![](https://i.imgur.com/Ks8s4cc.png)

<sup>Created on 2024-08-12 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by EvaMaeRey** [reply]
Helpful or distracting?

![image](https://github.com/user-attachments/assets/e23973c9-a7d2-4d8b-98d5-905f6c0419b1)

**Reply by larmarange** [reply]
Dear @EvaMaeRey 

Just to mention that there is a bug in CRAN version of `ggstats` when only one category is displayed (`position_likert()` fails to center the bar). It has been fixed in the dev version.

``` r
library(tidyverse)
library(ggstats) # dev version

age_cats <- c("18-29", "30-49"," 50-64", "65+")
sex_cats <- c("Male", "Female")

sample40replaceT <- function(x){
  sample(x, size = 40, replace = T) %>% factor(levels = x)
}

survey <- data.frame(c1 = sample40replaceT(age_cats),
                     c2 = sample40replaceT(sex_cats))

survey %>% 
  ggplot() + 
  aes(y = c1, fill = c2) + 
  geom_bar(position = "likert_count")
```

![](https://i.imgur.com/1pjrXvM.png)

``` r

# drop shadow layer (if MF counts were equal) attempt
last_plot() + 
  aes(fill = NULL)
```

![](https://i.imgur.com/4SLMPmk.png)

``` r

survey |> 
  ggplot() +
  aes(y = c1) +
  geom_bar(position = "likert_count") +
  geom_bar(
    mapping = aes(fill = c2),
    position = "likert_count"
  ) +
  theme_light()
```

![](https://i.imgur.com/qLGPHKn.png)

<sup>Created on 2024-08-14 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Reply by larmarange** [reply]
Just to mention that `ggstats` also provides a `stat_prop()` allowing to compute proportions according to a custom denominator. So you could use it to generate a pyramid with unequal size of sex groups.

``` r
library(tidyverse)
library(ggstats) # dev version

titanic <- Titanic |> 
  as_tibble() |> 
  tidyr::uncount(n)

titanic |> 
  ggplot() +
  aes(y = Class, fill = Sex) +
  geom_bar(position = "likert_count")
```

![](https://i.imgur.com/lWar3Bd.png)

``` r

titanic |> 
  ggplot() +
  aes(
    y = Class, fill = Sex,
    x = after_stat(prop),
    by = as.factor(Sex)
  ) +
  geom_bar(
    stat = "prop", # from ggstats
    position = "likert_count"
  ) +
  scale_x_continuous(labels = label_percent_abs())
```

![](https://i.imgur.com/cRpL3Kz.png)

<sup>Created on 2024-08-14 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Reply by EvaMaeRey** [reply]
I've tried stat_prop().  Seems very useful.

**Comment by EvaMaeRey** [comment]
Alternative, show surplus ![image](https://github.com/user-attachments/assets/287baf05-c079-4176-a12a-9aba63c0a7e9)

**Comment by larmarange** [comment]
The package `ggbreak` proposes a `scale_x_cut()` allowing to cut an axis. The current implementation does not allow to pass a custom label formater, nor to move the axis labels inside the cut. But maybe a scale could be a good idea to cut the x-axis and move axis.text inside the cut?

``` r
library(tidyverse)
library(ggstats) # dev version
library(ggbreak)
#> ggbreak v0.1.2
#> 
#> If you use ggbreak in published research, please cite the following
#> paper:
#> 
#> S Xu, M Chen, T Feng, L Zhan, L Zhou, G Yu. Use ggbreak to effectively
#> utilize plotting space to deal with large datasets and outliers.
#> Frontiers in Genetics. 2021, 12:774846. doi: 10.3389/fgene.2021.774846

titanic <- Titanic |> 
  as_tibble() |> 
  tidyr::uncount(n)


titanic |> 
  ggplot() +
  aes(
    y = Class, fill = Sex,
    x = after_stat(prop),
    by = as.factor(Sex)
  ) +
  geom_bar(
    stat = "prop", # from ggstats
    position = "likert_count"
  ) +
  scale_x_cut(
    breaks = 0,
    space = 1
  )
```

![](https://i.imgur.com/pj08YaT.png)

<sup>Created on 2024-08-14 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Reply by EvaMaeRey** [reply]
Super interesting.  Curious about the motivating use cases!

**Comment by EvaMaeRey** [comment]
Here is one more pyramid option... https://github.com/thomas-neitmann/ggcharts/blob/master/R/pyramid_chart.R

**Reply by EvaMaeRey** [reply]
![image](https://github.com/user-attachments/assets/e93e6237-5c52-4494-b07b-56869767178e)
